vars/piperPipelineStageCompliance.groovy

```diff
@@ -31,15 +34,33 @@ void call(Map parameters = [:]) {
         .mixinGeneralConfig(script.commonPipelineEnvironment, GENERAL_CONFIG_KEYS)
         .mixinStageConfig(script.commonPipelineEnvironment, stageName, STEP_CONFIG_KEYS)
         .mixin(parameters, PARAMETER_KEYS)
+        .addIfEmpty('sonarExecuteScan', script.commonPipelineEnvironment.configuration.runStep?.get(stageName)?.sonarExecuteScan)
         .use()
 
     piperStageWrapper (script: script, stageName: stageName) {
 
         // telemetry reporting
         utils.pushToSWA([step: STEP_NAME], config)
 
-        //ToDO: provide stage implementation
-        echo "${STEP_NAME}: Stage implementation is not provided yet. You can extend the stage using the provided stage extension mechanism."
-
+        if (config.sonarExecuteScan) {
+            durationMeasure(script: script, measurementName: 'sonar_duration') {
+                Map stepParams = [
+                    script: script
+                ]
+                if (!isProductiveBranch(script) && !isPullRequest()) {
```

Comment:
I think this logic should rather be added to the groovy wrapper of sonarExecuteScan if it is required or?

Reply:
You are making a good point.

Reply:
Turns out it already had special handling for PR branches, but not the handling for non-PR, non-productive-branch... so it fit right in. Writing a unit test for it turned out to be such a hassle that I decided to give up.